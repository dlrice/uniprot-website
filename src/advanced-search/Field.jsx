import React, { Fragment } from 'react';
import RangeField from './RangeField';
import EnumField from './EnumField';
import TextField from './TextField';
import AutocompleteWrapper from '../autocomplete/AutocompleteWrapper';

const Field = ({
  field, handleInputChange, handleRangeInputChange, queryInput,
}) => {
  const { dataType, hasRange } = field;
  let node;
  switch (dataType) {
    case 'enum':
      node = (
        <EnumField field={field} handleChange={handleInputChange} value={queryInput.stringValue} />
      );
      break;
    case 'date':
      node = <RangeField type="date" field={field} handleChange={handleRangeInputChange} />;
      break;
    case 'string':
      node = (
        <Fragment>
          {field.autoComplete ? (
            <AutocompleteWrapper
              url={field.autoComplete}
              onSelect={handleInputChange}
              title={field.label}
              value={queryInput.stringValue}
            />
          ) : (
            <TextField
              field={field}
              handleChange={handleInputChange}
              type="text"
              value={queryInput.stringValue}
            />
          )}
          {hasRange && dataType !== 'integer' && (
            <RangeField
              field={field}
              handleChange={handleRangeInputChange}
              type="text"
              rangeFrom={queryInput.rangeFrom}
              rangeTo={queryInput.rangeTo}
            />
          )}
        </Fragment>
      );
      break;
    case 'integer':
      if (hasRange) {
        node = RangeField({
          field,
          handleChange: handleRangeInputChange,
        });
      }
      node = TextField({
        field,
        type: 'number',
        handleChange: { handleInputChange },
      });
      break;
    default:
      return null;
  }
  return <div className="advanced-search__inputs">{node}</div>;
};

export default Field;
