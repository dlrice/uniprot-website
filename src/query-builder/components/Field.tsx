import React, { FC } from 'react';

import RangeField from './RangeField';
import EnumField from './EnumField';
import TextField from './TextField';
import EvidenceField from './EvidenceField';
import AutocompleteField from './AutocompleteField';

import {
  DataType,
  FieldType,
  QueryBit,
  SearchTermType,
} from '../types/searchTypes';

type FieldProps = {
  field: SearchTermType;
  handleChange: (updatedQueryBit: QueryBit) => void;
};

const Field: FC<FieldProps> = ({ field, handleChange }: FieldProps) => {
  const { dataType, fieldType, autoComplete } = field;
  let GenericField: FC<FieldProps & { initialValue?: QueryBit }>;

  switch (true) {
    case Boolean(autoComplete):
      GenericField = AutocompleteField;
      break;
    case dataType === DataType.enum || dataType === DataType.boolean:
      GenericField = EnumField;
      break;
    case dataType === DataType.date:
    case dataType === DataType.integer && fieldType === FieldType.range:
      GenericField = RangeField;
      break;
    case dataType === DataType.string && fieldType === FieldType.evidence:
      GenericField = EvidenceField;
      break;
    case dataType === DataType.string:
    case dataType === DataType.integer:
      GenericField = TextField;
      break;
    default:
      return null;
  }

  return (
    <GenericField
      field={field}
      handleChange={handleChange}
      // test
      // initialValue={{
      //   id_gene: '(gene:ydj1)',
      //   id_all: 'free text',
      // }}
    />
  );
};

export default Field;
