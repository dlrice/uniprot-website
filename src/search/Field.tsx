import React, { Fragment } from 'react';
import RangeField from './RangeField';
import EnumField from './EnumField';
import TextField from './TextField';
import AutocompleteWrapper from '../autocomplete/AutocompleteWrapper';
import { FieldType, Input } from './types/searchTypes';
import query from './state/reducers';

type FieldProps = {
  field: FieldType;
  handleInputChange: (clauseId: string, value: string | number) => void;
  handleRangeInputChange: (value: number, from?: boolean) => void;
  queryInput: Input;
};

const Field = ({
  field, handleInputChange, handleRangeInputChange, queryInput,
}: FieldProps) => {
  const { dataType, hasRange } = field;
  let node;
  switch (dataType) {
    case 'enum':
      node = (
        <EnumField
          field={field}
          handleChange={handleInputChange}
          value={queryInput ? queryInput.stringValue : ''}
        />
      );
      break;
    case 'date':
      node = <RangeField type="date" field={field} handleChange={handleRangeInputChange} />;
      break;
    case 'string':
      node = (
        <Fragment>
          {field.autoComplete ? (
            <AutocompleteWrapper
              url={field.autoComplete}
              onSelect={handleInputChange}
              title={field.label}
              value={queryInput.stringValue}
            />
          ) : (
            <TextField
              field={field}
              handleChange={handleInputChange}
              type="text"
              value={queryInput.stringValue}
            />
          )}
          {hasRange && dataType !== 'integer' && (
            <RangeField
              field={field}
              handleChange={handleRangeInputChange}
              type="text"
              rangeFrom={queryInput.rangeFrom}
              rangeTo={queryInput.rangeTo}
            />
          )}
        </Fragment>
      );
      break;
    case 'integer':
      if (hasRange) {
        return RangeField({
          field,
          handleChange: handleRangeInputChange,
          rangeFrom: queryInput.rangeFrom,
          rangeTo: queryInput.rangeTo,
        });
      }
      return TextField({
        field,
        type: 'number',
        handleChange: handleInputChange,
        value: queryInput.stringValue,
      });
      break;
    default:
      return null;
  }
  return <div className="advanced-search__inputs">{node}</div>;
};

export default Field;
