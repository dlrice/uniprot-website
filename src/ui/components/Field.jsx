import RangeField from './RangeField';
import EnumField from './EnumField';
import TextField from './TextField';

const dataTypes = { string: 'text', integer: 'number' };

const Field = (field, handleInputChange, handleRangeInputChange) => {
  const { dataType, hasRange } = field;
  if (dataType === 'enum') {
    return EnumField({ field, handleInputChange });
  }
  if (dataType === 'date') {
    return RangeField({ field, handleRangeInputChange });
  }
  if (!hasRange || dataType !== 'integer') {
    return TextField({
      field,
      handleInputChange,
      type: dataTypes[dataType],
    });
  }
  if (hasRange) {
    return RangeField({
      field,
      handleRangeInputChange,
      type: 'number',
    });
  }
  return null;
};

export default Field;
