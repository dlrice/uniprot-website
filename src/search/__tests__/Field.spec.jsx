import React from 'react';
import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Field from '../Field';

configure({ adapter: new Adapter() });
const handleInputChange = jest.fn();
const handleRangeInputChange = jest.fn();

describe('Clause component', () => {
  test('should render a `text` field', () => {
    const field = {
      label: 'UniProtKB AC',
      itemType: 'single',
      term: 'accession',
      dataType: 'string',
      description: 'Search by UniProtKB Accession',
      example: 'P12345',
    };
    const component = shallow(
      <Field
        field={field}
        handleInputChange={handleInputChange}
        handleRangeInputChange={handleRangeInputChange}
        queryInput={{}}
      />,
    );
    expect(component.debug()).toMatchSnapshot();
  });

  test('should render a `string` field with range', () => {
    const field = {
      label: 'UniProtKB AC',
      itemType: 'single',
      term: 'accession',
      dataType: 'string',
      hasRange: true,
      description: 'Search by UniProtKB Accession',
      example: 'P12345',
    };
    const component = shallow(
      <Field
        field={field}
        handleInputChange={handleInputChange}
        handleRangeInputChange={handleRangeInputChange}
        queryInput={{}}
      />,
    );
    expect(component.debug()).toMatchSnapshot();
  });

  test('should render an `enum` field', () => {
    const field = {
      dataType: 'enum',
    };
    const component = shallow(<Field field={field} />);
    expect(component.debug()).toMatchSnapshot();
  });

  test('should render an `integer` field', () => {
    const field = {
      label: 'Any',
      itemType: 'feature',
      term: 'sites',
      dataType: 'integer',
      description: 'Search by feature sites',
      example: 'translocation',
    };
    const component = shallow(
      <Field
        field={field}
        handleInputChange={handleInputChange}
        handleRangeInputChange={handleRangeInputChange}
        queryInput={{}}
      />,
    );
    expect(component.debug()).toMatchSnapshot();
  });

  test('should render an `integer` `range` field', () => {
    const field = {
      label: 'Any',
      itemType: 'feature',
      term: 'sites',
      dataType: 'integer',
      hasRange: true,
      description: 'Search by feature sites',
      example: 'translocation',
    };
    const component = shallow(
      <Field
        field={field}
        handleInputChange={handleInputChange}
        handleRangeInputChange={handleRangeInputChange}
        queryInput={{}}
      />,
    );
    expect(component.debug()).toMatchSnapshot();
  });

  test('should render `date` field', () => {
    const field = {
      label: 'Date Of Creation',
      itemType: 'single',
      term: 'created',
      dataType: 'date',
      hasRange: true,
      description: 'Search by Date of creation',
      example: '[2018-03-04 TO 2018-03-08]',
    };
    const component = shallow(
      <Field
        field={field}
        handleInputChange={handleInputChange}
        handleRangeInputChange={handleRangeInputChange}
        queryInput={{}}
      />,
    );
    expect(component.debug()).toMatchSnapshot();
  });

  test('should not return anything', () => {
    const field = {
      label: 'I dont exist',
      dataType: 'whatever',
    };
    const component = shallow(
      <Field
        field={field}
        handleInputChange={handleInputChange}
        handleRangeInputChange={handleRangeInputChange}
        queryInput={{}}
      />,
    );
    expect(component.debug()).toMatchSnapshot();
  });
});
