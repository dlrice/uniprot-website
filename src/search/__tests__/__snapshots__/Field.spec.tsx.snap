// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Clause component should not return anything 1`] = `""`;

exports[`Clause component should render \`date\` field 1`] = `
"<div className=\\"advanced-search__inputs\\">
  <RangeField type=\\"date\\" field={{...}} handleChange={[Function: mockConstructor]} rangeFrom={[undefined]} rangeTo={[undefined]} />
</div>"
`;

exports[`Clause component should render a \`string\` field with range 1`] = `
"<div className=\\"advanced-search__inputs\\">
  <TextField field={{...}} handleChange={[Function: mockConstructor]} type=\\"text\\" value={[undefined]} />
  <RangeField field={{...}} handleChange={[Function: mockConstructor]} type=\\"text\\" rangeFrom={[undefined]} rangeTo={[undefined]} />
</div>"
`;

exports[`Clause component should render a \`text\` field 1`] = `
"<div className=\\"advanced-search__inputs\\">
  <TextField field={{...}} handleChange={[Function: mockConstructor]} type=\\"text\\" value={[undefined]} />
</div>"
`;

exports[`Clause component should render an \`enum\` field 1`] = `
"<div className=\\"advanced-search__inputs\\">
  <EnumField field={{...}} handleChange={[undefined]} value=\\"\\" />
</div>"
`;

exports[`Clause component should render an \`integer\` \`range\` field 1`] = `
"<Fragment>
  <label htmlFor=\\"from_input_undefined\\">
    From
    <input id=\\"from_input_undefined\\" type={[undefined]} onChange={[Function: onChange]} placeholder=\\"0\\" value=\\"\\" />
  </label>
  <label htmlFor=\\"to_input_undefined\\">
    To
    <input id=\\"to_input_undefined\\" type={[undefined]} onChange={[Function: onChange]} placeholder=\\"100\\" value=\\"\\" />
  </label>
</Fragment>"
`;

exports[`Clause component should render an \`integer\` field 1`] = `
"<div className=\\"advanced-search__inputs\\">
  <label htmlFor=\\"input_sites\\">
    Any
    <input id=\\"input_sites\\" type=\\"number\\" value=\\"\\" onChange={[Function: onChange]} placeholder=\\"translocation\\" />
  </label>
</div>"
`;
